### PR TITLE
Large message v4

### PR DESCRIPTION
**Pull request by juimonen**

### CODE REVIEW

**Comment by juimonen** [comment]
related firmware change https://github.com/thesofproject/sof/pull/1134

**Comment by juimonen** [comment]
@lyakh can you check the ipc lock/unlock so first patch...

File: sound/soc/sof/ipc.c [hunk 1]
```diff
@@ -266,7 +257,28 @@ int sof_ipc_tx_message(struct snd_sof_ipc *ipc, u32 header,
...
+	struct snd_sof_dev *sdev = ipc->sdev;
+	struct snd_sof_ipc_msg *msg;
+	int ret;
+
+	if (msg_bytes > SOF_IPC_MSG_MAX_SIZE ||
+	    reply_bytes > SOF_IPC_MSG_MAX_SIZE)
+		return -ENOBUFS;
+
+	/* Serialise IPC TX */
+	mutex_lock(&ipc->tx_mutex);
```
Comment (plbossart) [inline sound/soc/sof/ipc.c:294]: I don't think we are on the right track here. Introducing a mutex which will only be required by Intel implementations smack in the middle of the core parts mixes layers. This should either be abstracted out or we implement the large message support in the platform driver.

File: sound/soc/sof/ipc.c [hunk 2]
```diff
@@ -551,6 +561,99 @@ int snd_sof_ipc_stream_posn(struct snd_sof_dev *sdev,
...
+	if (send)
+		sof_get_ctrl_copy_params(cdata->type, cdata, partdata, sparams);
+	else
+		sof_get_ctrl_copy_params(cdata->type, partdata, cdata, sparams);
+
+	/* copy the header data */
+	memcpy(partdata, cdata, sparams->hdr_bytes);
+
+	/* Serialise IPC TX */
+	mutex_lock(&sdev->ipc->tx_mutex);
```
Comment (plbossart) [inline sound/soc/sof/ipc.c:633]: same here, this shouldn't be used here.

File: sound/soc/sof/ipc.c [hunk 3]
```diff
@@ -564,6 +667,9 @@ int snd_sof_ipc_set_get_comp_data(struct snd_sof_ipc *ipc,
 	struct snd_sof_dev *sdev = ipc->sdev;
 	struct sof_ipc_ctrl_data *cdata = scontrol->control_data;
 	size_t send_bytes;
+	struct sof_ipc_ctrl_data_params sparams;
+	struct sof_ipc_fw_ready *ready = &sdev->fw_ready;
+	struct sof_ipc_fw_version *v = &ready->version;
```
Comment (plbossart) [inline sound/soc/sof/ipc.c:681]: xmas-tree style. Not my choice, this is what kernel reviewers like.

File: sound/soc/sof/control.c [hunk 4]
```diff
@@ -498,13 +489,9 @@ int snd_sof_bytes_ext_get(struct snd_kcontrol *kcontrol,
 	data_size = cdata->data->size + sizeof(const struct sof_abi_hdr);
 
 	/* check size. min3() is not used to avoid sparse warnings */
-	if (size < max_size)
-		max_size = size;
-	if (be->max < max_size)
-		max_size = be->max;
-	if (data_size > max_size) {
+	if (data_size > be->max) {
```
Comment (plbossart) [inline sound/soc/sof/control.c:527]: the comment on min3() should be removed.
I am not sure what this change does, is this really required

File: sound/soc/sof/ipc.c (hunk 1 above)
Comment (juimonen) [reply sound/soc/sof/ipc.c:294]: @plbossart this mutex is there already by @lyakh. I 'm just making here "unlocked" version of it to use it in large messages. So this mutex use is in sof-dev integrated.

File: sound/soc/sof/ipc.c (hunk 2 above)
Comment (juimonen) [reply sound/soc/sof/ipc.c:633]: @plbossart I tried to the mimic the mutex use in sof-dev. So surrounding the whole loop with a mutex was the way to go I understood from @lyakh...

**Comment by juimonen** [comment]
@plbossart should I see something in the kernel build... there's like nothing

File: sound/soc/sof/control.c (hunk 4 above)
Comment (juimonen) [reply sound/soc/sof/control.c:527]: old comment which was already obsolete before this change, will reword

File: sound/soc/sof/ipc.c (hunk 3 above)
Comment (juimonen) [reply sound/soc/sof/ipc.c:681]: ok

Comment (lyakh) [reply sound/soc/sof/ipc.c:681]: but you cannot just swap these here - they're interdependent. I usually also loosely follow the tree pattern, although not too strictly, but with interdependencies I actually prefer an uglier tree to using explicit assignments below declarations. I.e. I'd rather do
```
struct x *y = arg;
struct longname *anotherlongname = y->member;
```
than
```
struct longname *anotherlongname;
struct x *y = arg;

anotherlongname = y->member;
```

File: sound/soc/sof/ipc.c [hunk 5]
```diff
@@ -549,6 +559,99 @@ int snd_sof_ipc_stream_posn(struct snd_sof_dev *sdev,
...
+
+	/* copy the payload data in a loop */
+	offset = 0;
+	for (i = 0; i < sparams->num_msg; i++) {
+		send_bytes = sparams->msg_bytes > sparams->pl_size ?
+			sparams->pl_size : sparams->msg_bytes;
+		partdata->num_elems = send_bytes;
+		partdata->rhdr.hdr.size = sparams->hdr_bytes + send_bytes;
+		partdata->msg_index = i;
+		sparams->msg_bytes -= send_bytes;
```
Comment (kv2019i) [inline sound/soc/sof/ipc.c]: Modifying the called struct is a bit dubious. I'd add a "remaining" local variable and decrement the bytes left from that and leave the sparams->msg_bytes intact.

File: sound/soc/sof/ipc.c (hunk 3 above)
Comment (juimonen) [reply sound/soc/sof/ipc.c:681]: @lyakh @plbossart modified to "about" xmas tree as @lyakh pointed out there are inter-dependencies. If you want it different, please be more specific how I should change it. The whole xmas ordering was new to me...

File: sound/soc/sof/ipc.c (hunk 5 above)
Comment (juimonen) [reply sound/soc/sof/ipc.c]: @kv2019i ok. I added another local variable also, so the ternary operation inside the loop fits one line and it should be nicer and easier to understand.

Comment (kv2019i) [reply sound/soc/sof/ipc.c]: Thanks! Looks good now.

File: sound/soc/sof/ipc.c [hunk 6]
```diff
@@ -549,6 +559,103 @@ int snd_sof_ipc_stream_posn(struct snd_sof_dev *sdev,
...
+
+	/* copy the header data */
+	memcpy(partdata, cdata, sparams->hdr_bytes);
+
+	/* Serialise IPC TX */
+	mutex_lock(&sdev->ipc->tx_mutex);
+
+	/* copy the payload data in a loop */
+	for (i = 0; i < sparams->num_msg; i++) {
+		send_bytes = msg_bytes > pl_size ? pl_size : msg_bytes;
```
Comment (lyakh) [inline sound/soc/sof/ipc.c]: please, use min() / min_t()

File: sound/soc/sof/ipc.c [hunk 7]
```diff
@@ -549,6 +559,103 @@ int snd_sof_ipc_stream_posn(struct snd_sof_dev *sdev,
...
+				       struct sof_ipc_ctrl_data *cdata,
+				       struct sof_ipc_ctrl_data_params *sparams,
+				       bool send)
+{
+	struct sof_ipc_ctrl_data *partdata;
+	u32 offset = 0;
+	u32 send_bytes;
+	u32 msg_bytes;
+	int err = 0;
+	u32 pl_size;
```
Comment (lyakh) [inline sound/soc/sof/ipc.c]: do all these parameters really have to be fixed width? I really think we're abusing those types. Please, use size_t, ssize_t, int, unsigned int, bool,... as appropriate.

File: sound/soc/sof/ipc.c (hunk 6 above)
Comment (juimonen) [reply sound/soc/sof/ipc.c]: ok

File: sound/soc/sof/ipc.c (hunk 7 above)
Comment (juimonen) [reply sound/soc/sof/ipc.c]: changed most to size_t

**Comment by lgirdwood** [comment]
@juimonen I guess there is another PR with only the ABI change ?

**Comment by juimonen** [comment]
@lgirdwood @plbossart abi change in https://github.com/thesofproject/linux/pull/738
This PR now has also the abi change as first commit.

**Comment by lgirdwood** [comment]
> @lgirdwood @plbossart abi change in #738
> This PR now has also the abi change as first commit.

Why would you have the ABI change in 2 PRs, this will conflict with #738.

**Comment by plbossart** [comment]
I'll merge now and fix conflicts if required.

**Pull request merged**